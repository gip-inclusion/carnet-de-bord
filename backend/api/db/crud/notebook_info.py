from uuid import UUID

from asyncpg import Record
from asyncpg.connection import Connection

from api.db.models.notebook_info import NotebookInfo, OrientationType


def parse_notebook_info(record: Record) -> NotebookInfo:
    return NotebookInfo.parse_obj(record)


async def insert_or_update_need_orientation(
    connection: Connection,
    notebook_id: UUID,
    orientation: OrientationType | None,
    need_orientation: bool,
) -> NotebookInfo | None:
    record = await connection.fetchrow(
        """
        INSERT INTO public.notebook_info (notebook_id, orientation, need_orientation)
        VALUES ($1, $2, $3)
        ON CONFLICT ON CONSTRAINT notebook_info_pkey
        DO UPDATE SET need_orientation = $3
        RETURNING notebook_id, orientation, need_orientation, created_at, updated_at
        """,
        notebook_id,
        orientation,
        need_orientation,
    )

    if record:
        return parse_notebook_info(record)


async def get_notebook_info(
    connection: Connection, notebook_id: UUID
) -> NotebookInfo | None:
    record = await connection.fetchrow(
        """
        SELECT * from public.notebook_info WHERE notebook=$1
        """,
        notebook_id,
    )

    if record:
        return NotebookInfo.parse_obj(record)
