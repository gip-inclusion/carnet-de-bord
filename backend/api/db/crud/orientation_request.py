import logging
from uuid import UUID

from asyncpg import Record
from asyncpg.connection import Connection

from api.db.models.orientation_request import OrientationRequest

logger = logging.getLogger(__name__)


async def get_orientation_request_by_id(
    connection: Connection, orientation_request_id: UUID
) -> OrientationRequest | None:
    orientation_request: Record | None = await connection.fetchrow(
        "SELECT * FROM orientation_request WHERE id = $1", orientation_request_id
    )

    if orientation_request:
        return OrientationRequest.parse_obj(orientation_request)
