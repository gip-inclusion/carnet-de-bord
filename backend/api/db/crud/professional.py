import logging

from asyncpg import Record
from asyncpg.connection import Connection

from api.db.models.professional import Professional, ProfessionalInsert

logger = logging.getLogger(__name__)


async def get_professional_with_query(
    connection: Connection, query: str, *args
) -> Professional | None:
    async with connection.transaction():

        professional_record: Record | None = await connection.fetchrow(
            "SELECT public.professional.*, public.account.id as account_id "
            "FROM public.professional "
            "LEFT JOIN public.account "
            "ON public.account.professional_id = public.professional.id " + query,
            *args,
        )

        if professional_record:
            return Professional.parse_obj(professional_record)


async def get_professional_by_email(
    connection: Connection, email: str
) -> Professional | None:
    return await get_professional_with_query(
        connection,
        "WHERE LOWER(public.professional.email) = LOWER($1)",
        email,
    )


async def insert_professional(
    connection: Connection, professional_insert: ProfessionalInsert
) -> Professional | None:

    record = await connection.fetchrow(
        """
        INSERT INTO public.professional (structure_id, email, lastname, firstname, position, mobile_number)
        VALUES ($1, $2, $3, $4, $5, $6) returning id, structure_id, email, lastname, firstname, position, mobile_number, created_at, updated_at
        """,
        *professional_insert.dict().values(),
    )

    if record:
        return await parse_professional_from_record(record)


async def parse_professional_from_record(record: Record) -> Professional:
    return Professional.parse_obj(record)
