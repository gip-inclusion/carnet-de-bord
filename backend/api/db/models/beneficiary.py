from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, Field

from api.db.models.notebook import Notebook


class Beneficiary(BaseModel):
    id: UUID
    email: str | None
    lastname: str
    firstname: str
    caf_number: str | None
    pe_number: str | None
    postal_code: str | None
    city: str | None
    address1: str | None
    address2: str | None
    mobile_number: str | None
    date_of_birth: date
    place_of_birth: str | None
    deployment_id: UUID
    created_at: datetime
    updated_at: datetime
    internal_id: str | None
    notebook: Notebook | None
    # BRSA users may not have an account
    # (account is created on the  first login attempt)
    account_id: UUID | None


class BeneficiaryImport(BaseModel):
    si_id: str = Field(..., title="Identifiant du SI interne")
    firstname: str = Field(..., title="Prénom")
    lastname: str = Field(..., title="Nom")
    date_of_birth: date = Field(..., title="Date de naissance")
    place_of_birth: str | None = Field(None, title="Lieu de naissance")
    phone_number: str | None = Field(None, title="Numéro de téléphone")
    email: str | None = Field(None, title="Email")
    address1: str | None = Field(None, title="Adresse")
    address2: str | None = Field(None, title="Complément d'adresse")
    postal_code: str | None = Field(None, title="Code postal")
    city: str | None = Field(None, title="Ville")
    work_situation: str | None = Field(None, title="Situation professionnelle")
    caf_number: str | None = Field(None, title="Numéro CAF/MSA")
    pe_number: str | None = Field(None, title="Identifiant Pole Emploi")
    right_rsa: str | None = Field(None, title="Droits RSA")
    right_are: bool | None = Field(None, title="Droits ARE")
    right_ass: bool | None = Field(None, title="Droits ASS")
    right_bonus: bool | None = Field(None, title="Droits Bonus")
    right_rqth: bool | None = Field(None, title="Droits RQTH")
    geographical_area: str | None = Field(None, title="Zone de mobilité")
    rome_code_description: str | None = Field(None, title="Emploi recherché")
    education_level: str | None = Field(None, title="Niveau de formation")
    structure_name: str | None = Field(None, title="Structure d'accompagnement")
    advisor_email: str | None = Field(None, title="Accompagnateur référent")
