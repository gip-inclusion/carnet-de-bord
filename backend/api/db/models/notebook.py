from datetime import datetime
from typing import List
from uuid import UUID

from pydantic import BaseModel
from strenum import StrEnum

from api.db.models.wanted_job import WantedJob


class Notebook(BaseModel):
    id: UUID
    beneficiary_id: UUID
    wanted_jobs: List[WantedJob]
    # @TODO: add other fields


class NotebookMemberInsert(BaseModel):
    notebook_id: UUID
    account_id: UUID
    last_visited_at: datetime | None = None
    member_type: str
    last_modified_at: datetime | None = None
    creator_id: datetime | None = None
    invitation_sent_at: datetime | None = None
    active: bool | None = None


class NotebookMember(NotebookMemberInsert):
    id: UUID
    created_at: datetime


class NotebookMemberTypeEnum(StrEnum):
    NO_REFERENT = "no_referent"
    REFERENT = "referent"
