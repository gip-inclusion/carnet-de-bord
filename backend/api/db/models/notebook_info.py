from datetime import datetime
from uuid import UUID

from pydantic import BaseModel
from strenum import StrEnum

from api.db.models.orientation_type import OrientationType


class NotebookInfo(BaseModel):
    notebook_id: UUID
    orientation: OrientationType | None
    need_orientation: bool = False
    created_at: datetime
    updated_at: datetime
