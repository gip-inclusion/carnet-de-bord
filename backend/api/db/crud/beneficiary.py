from datetime import date
from typing import List
from uuid import UUID

from asyncpg import Record
from asyncpg.connection import Connection
from asyncpg.exceptions import UniqueViolationError

from api.db.crud.notebook import (
    NOTEBOOK_BASE_FIELDS,
    NOTEBOOK_BASE_JOINS,
    parse_notebook_from_record,
)
from api.db.models.beneficiary import Beneficiary, BeneficiaryImport

BENEFICIARY_BASE_QUERY = (
    """SELECT b.*, acc.id as account_id,"""
    + NOTEBOOK_BASE_FIELDS
    + """
    FROM public.beneficiary b
    LEFT JOIN notebook n ON n.beneficiary_id = b.id
    """
    + NOTEBOOK_BASE_JOINS
    + """
    LEFT JOIN public.account acc
    ON acc.beneficiary_id = b.id
"""
)


# see https://betakuang.medium.com/why-postgresqls-on-conflict-cannot-find-my-partial-unique-index-552327b85e1
firstname_lastname_date_of_birth_unique_idx = (
    "(LOWER(trim(firstname)), LOWER(trim(lastname)), date_of_birth, deployment_id)"
)


async def select_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    return await connection.fetch(
        f"""
SELECT (firstname, lastname, date_of_birth, internal_id, deployment_id, id)
FROM beneficiary
WHERE (firstname = $1 AND lastname = $2 AND date_of_birth = $3)
OR (internal_id = $4 AND deployment_id = $5)
        """,
        beneficiary.firstname,
        beneficiary.lastname,
        beneficiary.date_of_birth,
        beneficiary.si_id,
        deployment_id,
    )


async def update_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    await connection.fetchrow(
        f"""
UPDATE beneficiary SET mobile_number = $1
where internal_id = $2
returning id
        """,
        beneficiary.phone_number,
        beneficiary.si_id,
    )


async def insert_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    await connection.fetchrow(
        f"""
INSERT INTO BENEFICIARY (firstname, lastname, internal_id, date_of_birth, deployment_id, mobile_number)
values ($1, $2, $3, $4, $5, $6)
returning id
        """,
        beneficiary.firstname,
        beneficiary.lastname,
        beneficiary.si_id,
        beneficiary.date_of_birth,
        deployment_id,
        beneficiary.phone_number,
    )


async def import_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    result = select_beneficiary(connection, beneficiary, deployment_id)
    existing_rows = await result
    if len(existing_rows) == 0:
        print("inserting new user")
        await insert_beneficiary(connection, beneficiary, deployment_id)
    else:
        print(existing_rows[0][0][4])
        if (
            len(existing_rows) == 1
            and existing_rows[0][0][0] == beneficiary.firstname
            and existing_rows[0][0][1] == beneficiary.lastname
            and existing_rows[0][0][2] == beneficiary.date_of_birth
            and existing_rows[0][0][3] == beneficiary.si_id
        ):
            # Doesn't work somehow
            #     and existing_rows[0][0][4] == deployment_id
            print("updating existing user")
            await update_beneficiary(connection, beneficiary, deployment_id)
        else:
            print("data problem")


async def get_beneficiary_with_query(
    connection: Connection, query: str, *args
) -> Beneficiary | None:

    async with connection.transaction():

        beneficiary = None

        beneficiary_records: List[Record] = await connection.fetch(
            BENEFICIARY_BASE_QUERY + query,
            *args,
        )

        for beneficiary_record in beneficiary_records:

            if beneficiary is None:
                beneficiary = Beneficiary.parse_obj(beneficiary_record)

            if beneficiary_record["n_id"] is not None:
                beneficiary.notebook = await parse_notebook_from_record(
                    beneficiary_record,
                    notebook=beneficiary.notebook,
                )

        return beneficiary


async def get_beneficiary_from_personal_information(
    connection: Connection, firstname: str, lastname: str, birth_date: date
) -> Beneficiary | None:

    return await get_beneficiary_with_query(
        connection,
        # We use LOWER(BTRIM(field)) here to match the index so postgres
        # can use the index rather than scanning the whole table
        "WHERE LOWER(BTRIM(b.firstname)) = LOWER($1) AND "
        "LOWER(BTRIM(b.lastname)) = LOWER($2) AND "
        "b.date_of_birth = $3",
        firstname,
        lastname,
        birth_date,
    )


async def get_beneficiary_by_id(
    connection: Connection, beneficiary_id: UUID
) -> Beneficiary | None:

    return await get_beneficiary_with_query(
        connection, "WHERE b.id = $1", beneficiary_id
    )
