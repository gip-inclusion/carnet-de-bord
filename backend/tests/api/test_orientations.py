from unittest import mock

from asyncpg.connection import Connection
from fastapi.testclient import TestClient

from cdb.api.db.crud.beneficiary import get_structures_for_beneficiary
from cdb.api.db.crud.notebook import get_notebook_members_by_notebook_id
from cdb.api.db.crud.orientation_request import get_orientation_request_by_id
from cdb.api.db.models.beneficiary import Beneficiary
from cdb.api.db.models.notebook import Notebook
from cdb.api.db.models.notebook_info import OrientationType
from cdb.api.db.models.orientation_request import OrientationRequest
from cdb.api.db.models.professional import Professional
from tests.utils.assert_helpers import assert_member, assert_structure

UPDATE_ORIENTATION_ENDPOINT_PATH = "/v1/orientations/change"


@mock.patch("cdb.api.core.emails.send_mail")
async def test_verify_no_token(
    _: mock.Mock,
    test_client: TestClient,
    professional_pierre_chevalier: Professional,
    notebook_sophie_tifour: Notebook,
):
    response = test_client.post(
        UPDATE_ORIENTATION_ENDPOINT_PATH,
        json={
            "orientation_type": OrientationType.social,
            "notebook_id": str(notebook_sophie_tifour.id),
            "structure_id": str(professional_pierre_chevalier.structure_id),
            "new_referent_account_id": str(professional_pierre_chevalier.account_id),
        },
    )
    assert response.status_code == 401
    json = response.json()
    assert json["detail"] == "Missing credentials"


@mock.patch("cdb.api.core.emails.send_mail")
async def test_professional_not_allowed_to_change_orientation(
    _: mock.Mock,
    test_client: TestClient,
    professional_pierre_chevalier: Professional,
    notebook_sophie_tifour: Notebook,
    get_professional_jwt: str,
):
    response = test_client.post(
        UPDATE_ORIENTATION_ENDPOINT_PATH,
        json={
            "orientation_type": OrientationType.social,
            "notebook_id": str(notebook_sophie_tifour.id),
            "structure_id": str(professional_pierre_chevalier.structure_id),
            "new_referent_account_id": str(professional_pierre_chevalier.account_id),
        },
        headers={"jwt-token": get_professional_jwt},
    )
    assert response.status_code == 403
    json = response.json()
    assert json["detail"] == "Operation forbidden to the given role"


@mock.patch("cdb.api.core.emails.send_mail")
async def test_admin_structure_not_allowed_to_change_orientation(
    _: mock.Mock,
    test_client: TestClient,
    professional_pierre_chevalier: Professional,
    notebook_sophie_tifour: Notebook,
    get_admin_structure_jwt: str,
):
    response = test_client.post(
        UPDATE_ORIENTATION_ENDPOINT_PATH,
        json={
            "orientation_type": OrientationType.social,
            "notebook_id": str(notebook_sophie_tifour.id),
            "structure_id": str(professional_pierre_chevalier.structure_id),
            "new_referent_account_id": str(professional_pierre_chevalier.account_id),
        },
        headers={"jwt-token": get_admin_structure_jwt},
    )
    assert response.status_code == 403
    json = response.json()
    assert json["detail"] == "Operation forbidden to the given role"


@mock.patch("cdb.api.core.emails.send_mail")
async def test_change_orientation_while_keeping_same_referent(
    _: mock.Mock,
    test_client: TestClient,
    professional_pierre_chevalier: Professional,
    notebook_sophie_tifour: Notebook,
    giulia_diaby_jwt: str,
    db_connection: Connection,
):
    response = test_client.post(
        UPDATE_ORIENTATION_ENDPOINT_PATH,
        json={
            "orientation_type": OrientationType.social,
            "notebook_id": str(notebook_sophie_tifour.id),
            "structure_id": str(professional_pierre_chevalier.structure_id),
            "new_referent_account_id": str(professional_pierre_chevalier.account_id),
        },
        headers={"jwt-token": giulia_diaby_jwt},
    )
    assert response.status_code == 200
    members = await get_notebook_members_by_notebook_id(
        db_connection,
        notebook_sophie_tifour.id,
    )

    # Check that no new member row was added since referent doesn't change
    assert_member(members, professional_pierre_chevalier)


@mock.patch("cdb.api.core.emails.send_mail")
async def test_change_orientation_assign_to_structure_not_referent(
    _: mock.Mock,
    test_client: TestClient,
    professional_paul_camara: Professional,
    beneficiary_sophie_tifour: Beneficiary,
    notebook_sophie_tifour: Notebook,
    giulia_diaby_jwt: str,
    db_connection: Connection,
):
    response = test_client.post(
        UPDATE_ORIENTATION_ENDPOINT_PATH,
        json={
            "orientation_type": OrientationType.social,
            "notebook_id": str(notebook_sophie_tifour.id),
            "structure_id": str(professional_paul_camara.structure_id),
            "new_referent_account_id": None,
        },
        headers={"jwt-token": giulia_diaby_jwt},
    )
    assert response.status_code == 200
    members = await get_notebook_members_by_notebook_id(
        db_connection,
        notebook_sophie_tifour.id,
    )

    # Check that former referent is no longer active
    assert (
        len(
            [
                member
                for member in members
                if member.member_type == "referent" and member.active
            ]
        )
        == 0
    )
    structures = await get_structures_for_beneficiary(
        db_connection,
        beneficiary_sophie_tifour.id,
    )
    # Check that former structure has been set as 'outdated'
    assert_structure(
        structures,
        beneficiary_status="outdated",
        structure_name="Centre Communal d'action social Livry-Gargan",
    )

    # Check that new structure has been added as 'current'
    assert_structure(
        structures,
        beneficiary_status="current",
        structure_name="Service Social Départemental",
    )


@mock.patch("cdb.api.core.emails.send_mail")
async def test_change_orientation_with_new_referent(
    _: mock.Mock,
    test_client: TestClient,
    professional_paul_camara: Professional,
    professional_pierre_chevalier: Professional,
    beneficiary_sophie_tifour: Beneficiary,
    notebook_sophie_tifour: Notebook,
    giulia_diaby_jwt: str,
    db_connection: Connection,
):
    response = test_client.post(
        UPDATE_ORIENTATION_ENDPOINT_PATH,
        json={
            "orientation_type": OrientationType.social,
            "notebook_id": str(notebook_sophie_tifour.id),
            "structure_id": str(professional_paul_camara.structure_id),
            "new_referent_account_id": str(professional_paul_camara.account_id),
        },
        headers={"jwt-token": giulia_diaby_jwt},
    )
    assert response.status_code == 200
    members = await get_notebook_members_by_notebook_id(
        db_connection,
        notebook_sophie_tifour.id,
    )

    # Check that former referent is no longer active
    assert_member(
        members, professional_pierre_chevalier, member_type="referent", active=False
    )

    # Check that former referent remains in the notebook as simple member
    assert_member(
        members, professional_pierre_chevalier, member_type="no_referent", active=True
    )

    # Check that new referent is in the notebook as referent
    assert_member(
        members, professional_paul_camara, member_type="referent", active=True
    )

    structures = await get_structures_for_beneficiary(
        db_connection,
        beneficiary_sophie_tifour.id,
    )
    # Check that former structure has been set as 'outdated'
    assert_structure(
        structures,
        beneficiary_status="outdated",
        structure_name="Centre Communal d'action social Livry-Gargan",
    )

    # Check that new structure has been added as 'current'
    assert_structure(
        structures,
        beneficiary_status="current",
        structure_name="Service Social Départemental",
    )


@mock.patch("cdb.api.core.emails.send_mail")
async def test_change_orientation_with_new_referent_when_beneficiary_has_no_structure(
    _: mock.Mock,
    test_client: TestClient,
    professional_paul_camara: Professional,
    beneficiary_sophie_tifour: Beneficiary,
    notebook_craig_reilly: Notebook,
    samy_rouate_jwt: str,
    db_connection: Connection,
):
    response = test_client.post(
        UPDATE_ORIENTATION_ENDPOINT_PATH,
        json={
            "orientation_type": OrientationType.social,
            "notebook_id": str(notebook_craig_reilly.id),
            "structure_id": str(professional_paul_camara.structure_id),
            "new_referent_account_id": str(professional_paul_camara.account_id),
        },
        headers={"jwt-token": samy_rouate_jwt},
    )
    assert response.status_code == 200
    members = await get_notebook_members_by_notebook_id(
        db_connection,
        notebook_craig_reilly.id,
    )

    # Check that new referent is in the notebook as referent
    assert_member(
        members, professional_paul_camara, member_type="referent", active=True
    )

    structures = await get_structures_for_beneficiary(
        db_connection,
        beneficiary_sophie_tifour.id,
    )

    # Check that new structure has been added as 'current'
    assert_structure(
        structures,
        beneficiary_status="current",
        structure_name="Centre Communal d'action social Livry-Gargan",
    )


@mock.patch("cdb.api.core.emails.send_mail")
async def test_change_orientation_with_orientation_request(
    _: mock.Mock,
    test_client: TestClient,
    professional_pierre_chevalier: Professional,
    notebook_jennings_dee: Notebook,
    orientation_request_jennings_dee: OrientationRequest,
    giulia_diaby_jwt: str,
    db_connection: Connection,
):
    response = test_client.post(
        UPDATE_ORIENTATION_ENDPOINT_PATH,
        json={
            "orientation_request_id": str(orientation_request_jennings_dee.id),
            "orientation_type": OrientationType.social,
            "notebook_id": str(notebook_jennings_dee.id),
            "structure_id": str(professional_pierre_chevalier.structure_id),
            "new_referent_account_id": str(professional_pierre_chevalier.account_id),
        },
        headers={"jwt-token": giulia_diaby_jwt},
    )

    orientation_request = await get_orientation_request_by_id(
        db_connection, orientation_request_jennings_dee.id
    )

    assert orientation_request is not None
    assert orientation_request.decided_at is not None
    assert orientation_request.decided_orientation_type_id == OrientationType.social
    assert orientation_request.status == "accepted"

    assert response.status_code == 200


@mock.patch("cdb.api.core.emails.send_mail")
async def test_send_email_to_members_with_orientation_request(
    mock_send_email: mock.Mock,
    snapshot,
    test_client: TestClient,
    professional_paul_camara: Professional,
    professional_edith_orial: Professional,
    notebook_jennings_dee: Notebook,
    orientation_request_jennings_dee: OrientationRequest,
    giulia_diaby_jwt: str,
):
    response = test_client.post(
        UPDATE_ORIENTATION_ENDPOINT_PATH,
        json={
            "orientation_type": OrientationType.social,
            "notebook_id": str(notebook_jennings_dee.id),
            "structure_id": str(professional_paul_camara.structure_id),
            "new_referent_account_id": str(professional_paul_camara.account_id),
            "orientation_request_id": str(orientation_request_jennings_dee.id),
        },
        headers={"jwt-token": giulia_diaby_jwt},
    )
    assert response.status_code == 200
    assert mock_send_email.call_count == 2

    former_email_referent = mock_send_email.call_args_list[0]
    assert former_email_referent.kwargs["to"] == professional_edith_orial.email
    assert former_email_referent.kwargs["subject"] == "Réorientation d’un bénéficiaire"

    email_new_referent = mock_send_email.call_args_list[1]
    assert email_new_referent.kwargs["to"] == professional_paul_camara.email
    assert email_new_referent.kwargs["subject"] == "Réorientation d’un bénéficiaire"

    assert snapshot == {
        "former_email_referent": former_email_referent.kwargs["message"],
        "email_new_referent": email_new_referent.kwargs["message"],
    }


@mock.patch("cdb.api.core.emails.send_mail")
async def test_send_email_to_members_without_orientation_request(
    mock_send_email: mock.Mock,
    snapshot,
    test_client: TestClient,
    professional_paul_camara: Professional,
    professional_pierre_chevalier: Professional,
    notebook_sophie_tifour: Notebook,
    giulia_diaby_jwt: str,
):
    response = test_client.post(
        UPDATE_ORIENTATION_ENDPOINT_PATH,
        json={
            "orientation_type": OrientationType.social,
            "notebook_id": str(notebook_sophie_tifour.id),
            "structure_id": str(professional_paul_camara.structure_id),
            "new_referent_account_id": str(professional_paul_camara.account_id),
        },
        headers={"jwt-token": giulia_diaby_jwt},
    )
    assert response.status_code == 200
    assert mock_send_email.call_count == 2

    former_email_referent = mock_send_email.call_args_list[0]
    assert former_email_referent.kwargs["to"] == professional_pierre_chevalier.email
    assert former_email_referent.kwargs["subject"] == "Réorientation d’un bénéficiaire"

    email_new_referent = mock_send_email.call_args_list[1]
    assert email_new_referent.kwargs["to"] == professional_paul_camara.email
    assert email_new_referent.kwargs["subject"] == "Réorientation d’un bénéficiaire"

    assert snapshot == {
        "former_email_referent": former_email_referent.kwargs["message"],
        "email_new_referent": email_new_referent.kwargs["message"],
    }


@mock.patch("cdb.api.core.emails.send_mail")
async def test_send_email_to_members_first_orientation(
    mock_send_email: mock.Mock,
    test_client: TestClient,
    snapshot,
    professional_paul_camara: Professional,
    notebook_noel_keller: Beneficiary,
    giulia_diaby_jwt: str,
):
    response = test_client.post(
        UPDATE_ORIENTATION_ENDPOINT_PATH,
        json={
            "orientation_type": OrientationType.social,
            "notebook_id": str(notebook_noel_keller.id),
            "structure_id": str(professional_paul_camara.structure_id),
            "new_referent_account_id": str(professional_paul_camara.account_id),
        },
        headers={"jwt-token": giulia_diaby_jwt},
    )
    assert response.status_code == 200
    assert mock_send_email.call_count == 1

    email_new_referent = mock_send_email.call_args_list[0]
    assert email_new_referent.kwargs["to"] == professional_paul_camara.email
    assert email_new_referent.kwargs["subject"] == "Orientation d’un bénéficiaire"
    assert snapshot == {"email_new_referent": email_new_referent.kwargs["message"]}


@mock.patch("cdb.api.core.emails.send_mail")
async def test_unconsistent_orientation_request(
    _: mock.Mock,
    test_client: TestClient,
    professional_paul_camara: Professional,
    notebook_sophie_tifour: Notebook,
    orientation_request_jennings_dee: OrientationRequest,
    giulia_diaby_jwt: str,
):
    response = test_client.post(
        UPDATE_ORIENTATION_ENDPOINT_PATH,
        json={
            "orientation_type": OrientationType.social,
            "notebook_id": str(notebook_sophie_tifour.id),
            "structure_id": str(professional_paul_camara.structure_id),
            "orientation_request_id": str(orientation_request_jennings_dee.id),
        },
        headers={"jwt-token": giulia_diaby_jwt},
    )

    assert response.status_code == 400
