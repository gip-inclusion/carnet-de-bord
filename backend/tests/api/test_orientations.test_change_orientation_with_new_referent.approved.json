{
    "creatorId": "2addd10f-9bd3-4d37-b3c9-10a6e2c4be4f",
    "event": {
        "orientation": "Social",
        "previousOrientation": "Professionnel",
        "previousReferent": "Pierre Chevalier",
        "previousStructure": "Centre Communal d'action social Livry-Gargan",
        "referent": "Paul Camara",
        "structure": "Service Social D\u00e9partemental"
    },
    "eventDate": "<date0>",
    "eventType": "orientation"
}
